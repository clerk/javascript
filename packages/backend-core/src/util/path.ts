const SEPARATOR = '/';
const MULTIPLE_SEPARATOR_REGEX = new RegExp(SEPARATOR + '{1,}', 'g');

type PathString = string | null | undefined;

export function joinPaths(...args: PathString[]): string {
  return args
    .filter(p => p)
    .join(SEPARATOR)
    .replace(MULTIPLE_SEPARATOR_REGEX, SEPARATOR);
}
