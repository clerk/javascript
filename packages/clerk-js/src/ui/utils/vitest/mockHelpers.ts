// import { jest } from '@jest/globals';
import type { LoadedClerk } from '@clerk/types';
import type { ActiveSessionResource } from '@clerk/types';
import { type Mocked, vi } from 'vitest';

import type { RouteContextValue } from '../../router';

type FunctionLike = (...args: any) => any;

type DeepVitestMocked<T> = T extends FunctionLike
  ? Mocked<T>
  : T extends object
    ? {
        [k in keyof T]: DeepVitestMocked<T[k]>;
      }
    : T;

// Removing jest.Mock type for now, relying on inference
type MockMap<T = any> = {
  [K in { [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never }[keyof T]]?: ReturnType<typeof vi.fn>;
};

const mockProp = <T>(obj: T, k: keyof T, mocks?: MockMap<T>) => {
  if (typeof obj[k] === 'function') {
    const mockFn = mocks?.[k] ?? vi.fn();
    (obj[k] as unknown as ReturnType<typeof vi.fn>) = mockFn;
  }
};

const mockMethodsOf = <T extends Record<string, any> | null = any>(
  obj: T,
  options?: {
    exclude: (keyof T)[];
    mocks: MockMap<T>;
  },
) => {
  if (!obj) {
    return;
  }
  Object.keys(obj)
    .filter(key => !options?.exclude.includes(key as keyof T))
    // Pass the specific MockMap for the object T being mocked
    .forEach(k => mockProp(obj, k, options?.mocks));
};

export const mockClerkMethods = (clerk: LoadedClerk): DeepVitestMocked<LoadedClerk> => {
  // Cast clerk to any to allow mocking properties
  const clerkAny = clerk as any;

  mockMethodsOf(clerkAny);
  if (clerkAny.client) {
    mockMethodsOf(clerkAny.client.signIn);
    mockMethodsOf(clerkAny.client.signUp);
    clerkAny.client.sessions?.forEach((session: ActiveSessionResource) => {
      const sessionAny = session as any;
      mockMethodsOf(sessionAny, {
        exclude: ['checkAuthorization'],
        mocks: {
          // Ensure touch mock matches expected signature if available, otherwise basic mock
          touch: vi.fn(() => Promise.resolve(session)),
        },
      });
      if (sessionAny.user) {
        mockMethodsOf(sessionAny.user);
        sessionAny.user.emailAddresses?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.phoneNumbers?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.externalAccounts?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.organizationMemberships?.forEach((m: any) => {
          mockMethodsOf(m);
          if (m.organization) {
            mockMethodsOf(m.organization);
          }
        });
        sessionAny.user.passkeys?.forEach((m: any) => mockMethodsOf(m));
      }
    });
  }
  mockProp(clerkAny, 'navigate');
  mockProp(clerkAny, 'setActive');
  mockProp(clerkAny, 'redirectWithAuth');
  mockProp(clerkAny, '__internal_navigateWithError');
  return clerkAny as DeepVitestMocked<LoadedClerk>;
};

export const mockRouteContextValue = ({ queryString = '' }: Partial<DeepVitestMocked<RouteContextValue>>) => {
  return {
    basePath: '',
    startPath: '',
    flowStartPath: '',
    fullPath: '',
    indexPath: '',
    currentPath: '',
    queryString,
    queryParams: {},
    getMatchData: vi.fn(),
    matches: vi.fn(),
    baseNavigate: vi.fn(),
    navigate: vi.fn(() => Promise.resolve(true)),
    resolve: vi.fn((to: string) => new URL(to, 'https://clerk.com')),
    refresh: vi.fn(),
    params: {},
  } as RouteContextValue; // Keep original type assertion, DeepVitestMocked applied to input only
};
