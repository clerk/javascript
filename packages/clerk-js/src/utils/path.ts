const SEPARATOR = '/';
const MULTIPLE_SEPARATOR_REGEX = new RegExp(SEPARATOR + '{1,}', 'g');

type PathString = string | null | undefined;

export function joinPaths(a: PathString, b: PathString): string {
  return [a, b]
    .filter(p => p)
    .join(SEPARATOR)
    .replace(MULTIPLE_SEPARATOR_REGEX, SEPARATOR);
}
