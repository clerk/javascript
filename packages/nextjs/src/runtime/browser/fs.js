const fs = undefined;
const path = undefined;

module.exports = { fs, path };
