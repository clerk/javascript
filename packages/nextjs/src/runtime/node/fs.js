const { existsSync, writeFileSync, readFileSync, appendFileSync, mkdirSync, rmSync } = require('node:fs');
const path = require('node:path');
const fs = {
  existsSync,
  writeFileSync,
  readFileSync,
  appendFileSync,
  mkdirSync,
  rmSync,
};

module.exports = { fs, path };
