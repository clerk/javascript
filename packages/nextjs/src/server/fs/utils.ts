/**
 * Attention: Only import this module when the node runtime is used.
 * We are using conditional imports to mitigate bundling issues with Next.js server actions on version prior to 14.1.0.
 */
// @ts-ignore
import nodeRuntime from '#safe-node-apis';

const throwMissingFsModule = (module: string) => {
  throw new Error(`Clerk: ${module} is missing. This is an internal error. Please contact Clerk's support.`);
};

const nodeFsOrThrow = () => {
  if (!nodeRuntime.fs) {
    throwMissingFsModule('fs');
  }
  return nodeRuntime.fs;
};

const nodePathOrThrow = () => {
  if (!nodeRuntime.path) {
    throwMissingFsModule('path');
  }
  return nodeRuntime.path;
};

const nodeCwdOrThrow = () => {
  if (!nodeRuntime.cwd) {
    throwMissingFsModule('cwd');
  }
  return nodeRuntime.cwd;
};

export { nodeCwdOrThrow, nodeFsOrThrow, nodePathOrThrow };
