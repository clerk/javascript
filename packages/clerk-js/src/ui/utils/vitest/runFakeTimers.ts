import { vi } from 'vitest';
// import { jest } from '@jest/globals';
import { act } from '@testing-library/react';

type WithAct = <T>(fn: T) => T;
const withAct = ((fn: any) =>
  (...args: any) => {
    act(() => {
      fn(...args);
    });
  }) as WithAct;

const advanceTimersByTime = withAct(vi.advanceTimersByTime.bind(vi));
const runAllTimers = withAct(vi.runAllTimers.bind(vi));
const runOnlyPendingTimers = withAct(vi.runOnlyPendingTimers.bind(vi));

const createFakeTimersHelpers = () => {
  return { advanceTimersByTime, runAllTimers, runOnlyPendingTimers };
};

type FakeTimersHelpers = ReturnType<typeof createFakeTimersHelpers>;
type RunFakeTimersCallback = (timers: FakeTimersHelpers) => void | Promise<void>;

export const runFakeTimers = <T extends RunFakeTimersCallback, R extends ReturnType<T>>(
  cb: T,
): R extends Promise<any> ? Promise<void> : void => {
  vi.useFakeTimers();
  const res = cb(createFakeTimersHelpers());
  if (res && 'then' in res) {
    // @ts-expect-error
    return res.finally(() => vi.useRealTimers());
  }
  vi.useRealTimers();
  // @ts-ignore
  return;
};
