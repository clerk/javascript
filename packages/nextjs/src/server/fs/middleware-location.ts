import { nodeCwdOrThrow, nodeFsOrThrow, nodePathOrThrow } from './utils';

function hasSrcAppDir() {
  const { existsSync } = nodeFsOrThrow();
  const path = nodePathOrThrow();
  const cwd = nodeCwdOrThrow();

  const projectWithAppSrc = path.join(cwd(), 'src', 'app');

  return !!existsSync(projectWithAppSrc);
}

function suggestMiddlewareLocation() {
  const fileExtensions = ['ts', 'js'] as const;
  const suggestionMessage = (
    extension: (typeof fileExtensions)[number],
    to: 'src/' | '',
    from: 'src/app/' | 'app/' | '',
  ) =>
    `Clerk: clerkMiddleware() was not run, your middleware file might be misplaced. Move your middleware file to ./${to}middleware.${extension}. Currently located at ./${from}middleware.${extension}`;

  const { existsSync } = nodeFsOrThrow();
  const path = nodePathOrThrow();
  const cwd = nodeCwdOrThrow();

  const projectWithAppSrcPath = path.join(cwd(), 'src', 'app');
  const projectWithAppPath = path.join(cwd(), 'app');

  const checkMiddlewareLocation = (
    basePath: string,
    to: 'src/' | '',
    from: 'src/app/' | 'app/' | '',
  ): string | undefined => {
    for (const fileExtension of fileExtensions) {
      if (existsSync(path.join(basePath, `middleware.${fileExtension}`))) {
        return suggestionMessage(fileExtension, to, from);
      }
    }
    return undefined;
  };

  if (existsSync(projectWithAppSrcPath)) {
    return (
      checkMiddlewareLocation(projectWithAppSrcPath, 'src/', 'src/app/') || checkMiddlewareLocation(cwd(), 'src/', '')
    );
  }

  if (existsSync(projectWithAppPath)) {
    return checkMiddlewareLocation(projectWithAppPath, '', 'app/');
  }

  return undefined;
}

export { suggestMiddlewareLocation, hasSrcAppDir };
