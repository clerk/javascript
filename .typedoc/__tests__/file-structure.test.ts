import { readdir } from 'fs/promises';
import { join, relative } from 'path';
import { describe, expect, it } from 'vitest';

// Same function as in custom-router.mjs
function toKebabCase(str: string) {
  return str.replace(/((?<=[a-z\d])[A-Z]|(?<=[A-Z\d])[A-Z](?=[a-z]))/g, '-$1').toLowerCase();
}

const OUTPUT_LOCATION = `${process.cwd()}/docs`;

async function scanDirectory(type: 'file' | 'directory' = 'file') {
  const arr: string[] = [];
  const dir = await readdir(OUTPUT_LOCATION, { recursive: true, withFileTypes: true });

  for (const entry of dir) {
    const relativePath = relative(OUTPUT_LOCATION, join(entry.parentPath, entry.name));

    if (type === 'file' && entry.isFile()) {
      arr.push(relativePath);
    } else if (type === 'directory' && entry.isDirectory()) {
      arr.push(relativePath);
    }
  }

  return arr;
}

describe('Typedoc output', () => {
  it('should only have these top-level folders', async () => {
    const folders = await scanDirectory('directory');

    expect(folders).toMatchInlineSnapshot(`
      [
        "clerk-react",
        "nextjs",
        "shared",
        "types",
      ]
    `);
  });
  it('should have a deliberate file structure', async () => {
    const files = await scanDirectory('file');

    expect(files).toMatchSnapshot();
  });
  it('should only contain lowercase files', async () => {
    const files = await scanDirectory('file');
    const upperCaseFiles = files.filter(file => /[A-Z]/.test(file));

    expect(upperCaseFiles).toHaveLength(0);
  });
  it('should only contain kebab-cased files', async () => {
    const files = await scanDirectory('file');
    const incorrectFiles = files.filter(file => file !== toKebabCase(file));

    expect(incorrectFiles).toHaveLength(0);
  });
  it('should not contain README files', async () => {
    const files = await scanDirectory('file');
    const readmeFiles = files.filter(file => file.includes('readme'));

    expect(readmeFiles).toHaveLength(0);
  });
});
