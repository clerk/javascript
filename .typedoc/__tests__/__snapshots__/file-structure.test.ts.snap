// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typedoc output > should have a deliberate file structure 1`] = `
[
  "types/act-claim.mdx",
  "types/act-jwt-claim.mdx",
  "types/active-session-resource.mdx",
  "types/add-payment-source-params.mdx",
  "types/cancel-subscription-params.mdx",
  "types/check-authorization-fn.mdx",
  "types/check-authorization-from-session-claims.mdx",
  "types/check-authorization-params-from-session-claims.mdx",
  "types/check-authorization-with-custom-permissions.mdx",
  "types/clerk-api-error.mdx",
  "types/clerk-api-response-error.mdx",
  "types/clerk-host-router.mdx",
  "types/clerk-jwt-claims.mdx",
  "types/clerk-paginated-response.mdx",
  "types/clerk-pagination-params.mdx",
  "types/clerk-pagination-request.mdx",
  "types/clerk-resource.mdx",
  "types/clerk-status.mdx",
  "types/clerk.mdx",
  "types/commerce-billing-namespace.mdx",
  "types/commerce-checkout-json.mdx",
  "types/commerce-checkout-resource.mdx",
  "types/commerce-checkout-totals-json.mdx",
  "types/commerce-checkout-totals.mdx",
  "types/commerce-feature-json.mdx",
  "types/commerce-feature-resource.mdx",
  "types/commerce-initialized-payment-source-json.mdx",
  "types/commerce-initialized-payment-source-resource.mdx",
  "types/commerce-money-json.mdx",
  "types/commerce-money.mdx",
  "types/commerce-payment-charge-type.mdx",
  "types/commerce-payment-json.mdx",
  "types/commerce-payment-resource.mdx",
  "types/commerce-payment-source-json.mdx",
  "types/commerce-payment-source-methods.mdx",
  "types/commerce-payment-source-resource.mdx",
  "types/commerce-payment-source-status.mdx",
  "types/commerce-payment-status.mdx",
  "types/commerce-plan-json.mdx",
  "types/commerce-plan-resource.mdx",
  "types/commerce-product-json.mdx",
  "types/commerce-product-resource.mdx",
  "types/commerce-statement-group-json.mdx",
  "types/commerce-statement-group.mdx",
  "types/commerce-statement-json.mdx",
  "types/commerce-statement-resource.mdx",
  "types/commerce-statement-status.mdx",
  "types/commerce-statement-totals-json.mdx",
  "types/commerce-statement-totals.mdx",
  "types/commerce-subscriber-type.mdx",
  "types/commerce-subscription-json.mdx",
  "types/commerce-subscription-plan-period.mdx",
  "types/commerce-subscription-resource.mdx",
  "types/commerce-subscription-status.mdx",
  "types/confirm-checkout-params.mdx",
  "types/create-checkout-params.mdx",
  "types/create-organization-params.mdx",
  "types/element-object-key.mdx",
  "types/elements-config.mdx",
  "types/get-payment-attempts-params.mdx",
  "types/get-payment-sources-params.mdx",
  "types/get-plans-params.mdx",
  "types/get-statements-params.mdx",
  "types/get-subscriptions-params.mdx",
  "types/get-token.mdx",
  "types/id-selectors.mdx",
  "types/initialize-payment-source-params.mdx",
  "types/internal_checkout-props.mdx",
  "types/internal_plan-details-props.mdx",
  "types/jwt-claims.mdx",
  "types/jwt-header.mdx",
  "types/legacy-redirect-props.mdx",
  "types/localization-resource.mdx",
  "types/make-default-payment-source-params.mdx",
  "types/multi-domain-and-or-proxy.mdx",
  "types/organization-custom-role-key.mdx",
  "types/organization-domain-resource.mdx",
  "types/organization-domain-verification-status.mdx",
  "types/organization-enrollment-mode.mdx",
  "types/organization-invitation-resource.mdx",
  "types/organization-invitation-status.mdx",
  "types/organization-membership-request-resource.mdx",
  "types/organization-membership-resource.mdx",
  "types/organization-permission-key.mdx",
  "types/organization-resource.mdx",
  "types/organization-suggestion-resource.mdx",
  "types/organization-suggestion-status.mdx",
  "types/organizations-jwt-claim.mdx",
  "types/override.mdx",
  "types/path-value.mdx",
  "types/payment-gateway.mdx",
  "types/pending-session-options.mdx",
  "types/pending-session-resource.mdx",
  "types/protect-props.mdx",
  "types/record-to-path.mdx",
  "types/redirect-options.mdx",
  "types/remove-payment-source-params.mdx",
  "types/reverification-config.mdx",
  "types/saml-strategy.mdx",
  "types/sdk-metadata.mdx",
  "types/server-get-token-options.mdx",
  "types/server-get-token.mdx",
  "types/session-resource.mdx",
  "types/session-status-claim.mdx",
  "types/session-verification-level.mdx",
  "types/session-verification-types.mdx",
  "types/set-active-params.mdx",
  "types/set-active.mdx",
  "types/sign-in-resource.mdx",
  "types/sign-out.mdx",
  "types/sign-up-authenticate-with-metamask-params.mdx",
  "types/sign-up-resource.mdx",
  "types/signed-in-session-resource.mdx",
  "types/state-selectors.mdx",
  "types/use-auth-return.mdx",
  "types/use-session-list-return.mdx",
  "types/use-session-return.mdx",
  "types/use-sign-in-return.mdx",
  "types/use-sign-up-return.mdx",
  "types/use-user-return.mdx",
  "types/user-organization-invitation-resource.mdx",
  "types/user-resource.mdx",
  "types/without.mdx",
  "shared/api-url-from-publishable-key.mdx",
  "shared/build-clerk-js-script-attributes.mdx",
  "shared/build-publishable-key.mdx",
  "shared/camel-to-snake.mdx",
  "shared/clerk-js-script-url.mdx",
  "shared/clerk-runtime-error.mdx",
  "shared/create-dev-or-staging-url-cache.mdx",
  "shared/create-path-matcher.mdx",
  "shared/deep-camel-to-snake.mdx",
  "shared/deep-snake-to-camel.mdx",
  "shared/deprecated-object-property.mdx",
  "shared/derive-state.mdx",
  "shared/extract-dev-browser-jwt-from-url.mdx",
  "shared/fast-deep-merge-and-replace.mdx",
  "shared/get-clerk-js-major-version-or-tag.mdx",
  "shared/get-cookie-suffix.mdx",
  "shared/get-env-variable.mdx",
  "shared/get-non-undefined-values.mdx",
  "shared/get-script-url.mdx",
  "shared/get-suffixed-cookie-name.mdx",
  "shared/icon-image-url.mdx",
  "shared/in-browser.mdx",
  "shared/is-browser-online.mdx",
  "shared/is-clerk-runtime-error.mdx",
  "shared/is-development-from-publishable-key.mdx",
  "shared/is-development-from-secret-key.mdx",
  "shared/is-ipv4-address.mdx",
  "shared/is-production-from-publishable-key.mdx",
  "shared/is-production-from-secret-key.mdx",
  "shared/is-publishable-key.mdx",
  "shared/is-staging.mdx",
  "shared/is-truthy.mdx",
  "shared/is-valid-browser-online.mdx",
  "shared/is-valid-browser.mdx",
  "shared/isomorphic-atob.mdx",
  "shared/load-clerk-js-script.mdx",
  "shared/pages-or-infinite-options.mdx",
  "shared/paginated-hook-config.mdx",
  "shared/paginated-resources.mdx",
  "shared/parse-publishable-key.mdx",
  "shared/read-json-file.mdx",
  "shared/set-clerk-js-loading-error-package-name.mdx",
  "shared/snake-to-camel.mdx",
  "shared/titleize.mdx",
  "shared/to-sentence.mdx",
  "shared/use-clerk.mdx",
  "shared/use-organization-list-params.mdx",
  "shared/use-organization-list-return.mdx",
  "shared/use-organization-list.mdx",
  "shared/use-organization-params.mdx",
  "shared/use-organization-return.mdx",
  "shared/use-organization.mdx",
  "shared/use-reverification.mdx",
  "shared/use-session-list.mdx",
  "shared/use-session.mdx",
  "shared/use-user.mdx",
  "shared/user-agent-is-robot.mdx",
  "shared/version-selector.mdx",
  "nextjs/auth.mdx",
  "nextjs/build-clerk-props.mdx",
  "nextjs/clerk-middleware-auth-object.mdx",
  "nextjs/clerk-middleware-options.mdx",
  "nextjs/clerk-middleware.mdx",
  "nextjs/create-async-get-auth.mdx",
  "nextjs/create-sync-get-auth.mdx",
  "nextjs/current-user.mdx",
  "nextjs/get-auth.mdx",
  "clerk-react/api-keys.mdx",
  "clerk-react/clerk-provider-props.mdx",
  "clerk-react/protect.mdx",
  "clerk-react/redirect-to-create-organization.mdx",
  "clerk-react/redirect-to-organization-profile.mdx",
  "clerk-react/redirect-to-user-profile.mdx",
  "clerk-react/use-auth.mdx",
  "clerk-react/use-clerk.mdx",
  "clerk-react/use-organization-list.mdx",
  "clerk-react/use-organization.mdx",
  "clerk-react/use-reverification.mdx",
  "clerk-react/use-session-list.mdx",
  "clerk-react/use-session.mdx",
  "clerk-react/use-sign-in.mdx",
  "clerk-react/use-sign-up.mdx",
  "clerk-react/use-user.mdx",
  "backend/allowlist-identifier.mdx",
  "backend/auth-object.mdx",
  "backend/authenticate-request-options.mdx",
  "backend/client.mdx",
  "backend/email-address.mdx",
  "backend/external-account.mdx",
  "backend/get-auth-fn.mdx",
  "backend/identification-link.mdx",
  "backend/infer-auth-object-from-token-array.mdx",
  "backend/infer-auth-object-from-token.mdx",
  "backend/invitation-status.mdx",
  "backend/invitation.mdx",
  "backend/organization-invitation-status.mdx",
  "backend/organization-invitation.mdx",
  "backend/organization-membership-public-user-data.mdx",
  "backend/organization-membership.mdx",
  "backend/organization-sync-options.mdx",
  "backend/organization-sync-target.mdx",
  "backend/organization.mdx",
  "backend/paginated-resource-response.mdx",
  "backend/phone-number.mdx",
  "backend/public-organization-data-json.mdx",
  "backend/redirect-url.mdx",
  "backend/saml-account.mdx",
  "backend/saml-connection.mdx",
  "backend/session-activity.mdx",
  "backend/session.mdx",
  "backend/user.mdx",
  "backend/verification.mdx",
  "backend/verify-machine-auth-token.mdx",
  "backend/verify-token-options.mdx",
  "backend/verify-token.mdx",
  "backend/verify-webhook-options.mdx",
  "backend/verify-webhook.mdx",
  "backend/web3-wallet.mdx",
]
`;
